src/cpp/src/whisper/whisper.cpp

```diff
@@ -0,0 +1,201 @@
+// Copyright (C) 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <iostream>
+#include <openvino/openvino.hpp>
+#include <regex>
+#include <thread>
+
+#include "../utils.hpp"
+#include "openvino/genai/streamer_base.hpp"
+#include "openvino/genai/whisper_generation_config.hpp"
+#include "openvino/genai/whisper_speech_recognition_pipeline.hpp"
+#include "whisper_feature_extractor.hpp"
+#include "whisper_models.hpp"
+
+namespace {
+
+void suppress_tokens(ov::Tensor& logits, const std::vector<int64_t>& suppress_tokens) {
+    auto data = logits.data<float>();
+    for (auto supress_token : suppress_tokens) {
+        data[supress_token] = -std::numeric_limits<float>::infinity();
+    }
+}
+
+ov::Tensor encode(ov::InferRequest& request, std::vector<float>& mel_data) {
+    ov::Shape input_shape = {1, 80, 3000};
+    ov::Tensor input_tensor(ov::element::f32, input_shape, mel_data.data());
+
+    request.set_tensor("input_features", input_tensor);
+
+    request.infer();
+
+    // reset input tensor
+    request.set_tensor("input_features", ov::Tensor(ov::element::f32, input_shape));
+
+    return request.get_tensor("last_hidden_state");
+}
+
+void set_past_kev_value(ov::InferRequest& source, ov::InferRequest& dest) {
+    // source outputs:
+    // present.0.decoder.key
+    // present.0.decoder.value
+    // present.0.encoder.key
+    // present.0.encoder.value
+
+    // dest imputs:
+    // past_key_values.0.decoder.key
+    // past_key_values.0.decoder.value
+    // past_key_values.0.encoder.key
+    // past_key_values.0.encoder.value
+
+    for (auto& dec_output : source.get_compiled_model().outputs()) {
+        std::string dec_output_name = dec_output.get_any_name();
+        if (dec_output_name.find("logits") != std::string::npos) {
+            continue;
+        }
+
+        std::string with_past_input_name =
+            std::regex_replace(dec_output_name, std::regex("present"), "past_key_values");
+
+        auto kv_tensor = source.get_tensor(dec_output_name);
```

Comment:
for GPU we can consider creation of RemoteTensors to avoid data copy from / to host.

Reply:
I took a look at `RemoteTensor` docs. I need more time to understand how to use them. Don't you mind is I'll address it in further PR? I added todo for this.

Reply:
yes, sure